### PR TITLE
Create middleware for verifying privilege token

### PR DESCRIPTION
**Pull request by 0xdev22**
# Proposed Changes

### CODE REVIEW

File: middleware/token_exchange/main.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		return fiber.NewError(fiber.StatusInternalServerError, err.Error())
+	}
+
+	tkID := c.Params("tokenID")
+
+	if tkID != claims.VehicleTokenID {
+		p.Log.Debug().Str("VehicleTokenID In Request", tkID).
+			Str("VehicleTokenID in bearer token", claims.VehicleTokenID).
+			Msg("Invalid vehicle token")
+		return fiber.NewError(fiber.StatusUnauthorized, "Unauthorized! Wrong vehicle token provided")
```
Comment (elffjs) [inline middleware/token_exchange/main.go:41]: This could be more specific too. Up to you whether it's worth the trouble.

File: middleware/token_exchange/main.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+func (p *PrivilegeHandler) HasTokenPrivilege(privilegeID int64) fiber.Handler {
+	return func(c *fiber.Ctx) error {
+		return p.checkPrivilege(c, privilegeID)
+	}
+}
+
+func (p *PrivilegeHandler) checkPrivilege(c *fiber.Ctx, privilegeID int64) error {
+	claims, err := getVehicleTokenClaims(c)
+	if err != nil {
+		return fiber.NewError(fiber.StatusInternalServerError, err.Error())
```
Comment (elffjs) [inline middleware/token_exchange/main.go:32]: We may not want to expose the inner error here to a user.

File: middleware/token_exchange/claims.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+type VehicleTokenClaims struct {
+	VehicleTokenID string
+	UserEthAddress string
+	Privileges     []int64
+}
+
+type vehicleTokenClaimsResponseRaw struct {
+	Sub        string
+	UserID     string
```
[the remark this answers is not in the document]
Comment (0xdev22) [reply middleware/token_exchange/claims.go:18]: I've created PR's to address this issue. Once they have been merged, I will pull them in and resolve this comment.

File: middleware/verify_privilege_token/claims.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package tokenexchange
+package verify_privilege_token
```
Comment (elffjs) [inline middleware/verify_privilege_token/claims.go:1]: I agree with the linter—you want to keep these short and free of punctuation. Honestly `tokenexchange` is good if that's what makes the most sense to you, we can always change it.

I think this is my only serious critique. The rest is details.

File: middleware/verify_privilege_token/claims.go [hunk 5]
```diff
@@ -36,7 +36,7 @@ func getVehicleTokenClaims(c *fiber.Ctx) (VehicleTokenClaims, error) {
 
 	return VehicleTokenClaims{
 		VehicleTokenID: p.Sub,
-		UserEthAddress: p.UserID,
-		Privileges:     p.Privileges,
+		UserEthAddress: p.UserEthAddress,
+		PrivilegeIDs:   p.Privileges,
```
Comment (elffjs) [inline middleware/verify_privilege_token/claims.go:40]: It seems a bit odd that "id" is on one side but not the other. If you want to stick to leaving it out everywhere I think that's okay.

File: middleware/verify_privilege_token/main.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (p *verifyPrivilegeToken) HasTokenPrivilege(privilegeID int64) fiber.Handler {
+	return func(c *fiber.Ctx) error {
+		return p.checkPrivilege(c, privilegeID)
+	}
+}
+
+func (p *verifyPrivilegeToken) checkPrivilege(c *fiber.Ctx, privilegeID int64) error {
+	logger := p.cfg.Log
```
Comment (elffjs) [inline middleware/verify_privilege_token/main.go:36]: We could log [the route](https://docs.gofiber.io/api/ctx#route) in this method. It's always a bit tricky to log in libraries, you could say that this isn't our job here. You can make a sub-logger with a fixed attribute by doing something like

```
logger := p.cfg.Log.With().Str("myAttr", "xdd").Logger()
```

Note that `logger` here is not a pointer for some reason. We might also log the user once we have it, just so we know who's having a bad time.

File: middleware/verify_privilege_token/main.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	claims, err := getVehicleTokenClaims(c)
+	if err != nil {
+		logger.Debug().Str("VehicleTokenID In Request", c.Params("tokenID")).
+			Str("VehicleTokenID in bearer token", claims.VehicleTokenID).
+			Msg(err.Error())
```
Comment (elffjs) [inline middleware/verify_privilege_token/main.go:41]:
```suggestion
		logger.Debug().Err(err).Str("pathTokenId", tkID).Msg("Couldn't parse JWT.")
```
After moving line 45 up a bit.

I don't think it makes sense to access `claims.VehicleTokenID` if there's an error, it seems like we always return a zero struct in that case.

I like to keep the `.Msg` part fixed so that it's easy to scan for the error in the log reader.

File: middleware/verify_privilege_token/main.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		return fiber.NewError(fiber.StatusInternalServerError, "Error verifying user privilege!")
+	}
+
+	tkID := c.Params("tokenID")
+
+	if tkID != claims.VehicleTokenID {
+		logger.Debug().Str("VehicleTokenID In Request", tkID).
+			Str("VehicleTokenID in bearer token", claims.VehicleTokenID).
+			Msg("Invalid vehicle token")
```
Comment (elffjs) [inline middleware/verify_privilege_token/main.go:50]:
```suggestion
		logger.Debug().Str("pathTokenId", tkID).
			Str("jwtTokenId", claims.VehicleTokenID).
			Msg("Mismatched token ids")
```
The exact wording isn't so important, but I think the gist is.

These are at such a low log level that we'll probably never see them.

File: middleware/verify_privilege_token/main.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		return fiber.NewError(fiber.StatusUnauthorized, "Unauthorized! Wrong vehicle token provided")
+	}
+
+	// Verify privilege is correct
+	privilegeFound := slices.Contains(claims.PrivilegeIDs, privilegeID)
+	if !privilegeFound {
+		logger.Debug().Interface("Privilege In Request", claims.PrivilegeIDs).Msg("Invalid privilege requested")
```
Comment (elffjs) [inline middleware/verify_privilege_token/main.go:57]:
```suggestion
		logger.Debug().Str("tokenId", tkID).Int64("requiredPrivilege", privilegeID).Interface("jwtPrivileges", claims.PrivilegeIDs).Msg("Token lacks the required privilege.")
```
Maybe a bit wordy. Not an exact suggestion.

File: middleware/verify_privilege_token/main.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"fmt"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/rs/zerolog"
+	"golang.org/x/exp/slices"
+)
+
+type VerifyPrivilegeTokenInterface interface {
```
Comment (elffjs) [inline middleware/verify_privilege_token/main.go:11]: Haha I think this is the same as `IVerifyPrivilegeToken` or whatever it was, we can go back to that, I think that's fine and it was a bit shorter.

File: middleware/verify_privilege_token/main.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (p *verifyPrivilegeToken) checkPrivilege(c *fiber.Ctx, privilegeID int64) error {
+	logger := p.cfg.Log
+	claims, err := getVehicleTokenClaims(c)
+	if err != nil {
+		logger.Debug().Str("VehicleTokenID In Request", c.Params("tokenID")).
+			Str("VehicleTokenID in bearer token", claims.VehicleTokenID).
+			Msg(err.Error())
+		return fiber.NewError(fiber.StatusInternalServerError, "Error verifying user privilege!")
```
Comment (elffjs) [inline middleware/verify_privilege_token/main.go:42]: This is probably a 401, yeah? What would you return if someone passed in a mangled identity token?

**Review by elffjs — CHANGES_REQUESTED** [review]

File: middleware/verify_privilege_token/main.go (hunk 10 above)
Comment (0xdev22) [reply middleware/verify_privilege_token/main.go:11]: lol.. kind of hard naming this in a way that reflect what it is specifically without the I or interface in the variable name

File: middleware/verify_privilege_token/main.go (hunk 11 above)
Comment (0xdev22) [reply middleware/verify_privilege_token/main.go:42]: hmmm.. i might need to look into how to get the specific error then, since this can also be caused (on rare occassions) by the server not being able to decode the jwt token.

File: middleware/privilegetoken/claims.go [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+type TokenClaims struct {
+	DeviceTokenID  string // Entity in this case is what we need privilege to access
+	UserEthAddress string
+	PrivilegeIDs   []int64
+}
+
+type privilegeTokenClaimsResponseRaw struct {
+	Sub            string
+	UserEthAddress string
+	PrivilegeIDs   []int64
```
Comment (elffjs) [inline middleware/privilegetoken/claims.go:19]: I think you're calling this `"privileges"` still in the token. That is fine, but one of these has to change.

**Review by elffjs — CHANGES_REQUESTED** [review]

**Review by elffjs — APPROVED** [review]
